desc/objectives/_coils.py

```diff
@@ -98,77 +98,48 @@ def build(self, use_jit=True, verbose=1):  # noqa:C901
         # local import to avoid circular import
         from desc.coils import CoilSet, MixedCoilSet
 
+        coil = self.things[0]
+
         self._dim_f = 0
         self._quad_weights = jnp.array([])
 
-        def to_list(coilset):
-            """Turn a MixedCoilSet container into a list of what it's containing."""
-            if isinstance(coilset, list):
-                return [to_list(x) for x in coilset]
-            elif isinstance(coilset, MixedCoilSet):
-                return [to_list(x) for x in coilset]
-            elif isinstance(coilset, CoilSet):
-                # use the same grid/transform for CoilSet
-                return to_list(coilset.coils[0])
-            else:
-                return [coilset]
-
-        # gives structure of coils, e.g. MixedCoilSet(coils, coils) would give a
-        # a structure of [[*, *], [*, *]] if n = 2 coils
-        coil_leaves, coil_structure = tree_flatten(
-            self.things[0], is_leaf=lambda x: not hasattr(x, "__len__")
-        )
-        self._num_coils = len(coil_leaves)
+        # get individual coils from coilset
+        coils = tree_leaves(coil, is_leaf=lambda x: hasattr(x, "_current"))
+        self._num_coils = len(coils)
 
-        # check type
+        # map grid to list with the same length as coils
         if isinstance(self._grid, numbers.Integral):
             self._grid = LinearGrid(N=self._grid, endpoint=False)
-        # all of these cases return a container MixedCoilSet that contains
-        # LinearGrids. i.e. MixedCoilSet.coils = list of LinearGrid
         if self._grid is None:
-            # map default grid to structure of inputted coils
-            self._grid = tree_map(
-                lambda x: LinearGrid(
-                    N=2 * x.N + 5, NFP=getattr(x, "NFP", 1), endpoint=False
-                ),
-                self.things[0],
-                is_leaf=lambda x: not hasattr(x, "__len__"),
-            )
+            self._grid = [LinearGrid(N=2 * c.N + 5, endpoint=False) for c in coils]
         elif isinstance(self._grid, _Grid):
-            # map inputted single LinearGrid to structure of inputted coils
             self._grid = [self._grid] * self._num_coils
-            self._grid = tree_unflatten(coil_structure, self._grid)
-        else:
-            # this case covers an inputted list of grids that matches the size
-            # of the inputted coils. Can be a 1D list or nested list.
-            flattened_grid = tree_leaves(
-                self._grid, is_leaf=lambda x: isinstance(x, _Grid)
-            )
-            self._grid = tree_unflatten(coil_structure, flattened_grid)
+        assert isinstance(self._grid, list)
+
+        self._dim_f = np.sum(
```

Comment:
can't this just be `sum(grid.num_nodes for grid in self._grid)`?

Reply:
we do basically the same for `quad_weights` below

Reply:
Actually I needed to update `quad_weights` to match the new logic like `dim_f`. Thanks for catching that. 

Before, the grids were mapped to the same tree structure as the input `coil`. That was nice because it always made sure the structures matched (there was a grid for every coil), but it was also awkward because we were creating a `MixedCoilSet` that had `Grid` elements instead of `Coils`. 

My changes now avoid that mapping, and the format for specifying the grids has changed a bit too. Say we have a `MixedCoilSet` that contains two `CoilSet`s, the first with 2 coils and the second with 3 coils. Before you would give the grid for each individual `Coil` like this: 
```
grid = [[grid1, grid2], [grid3, grid4, grid5]]
```
And now you only give the grids for each `CoilSet` like this: 
```
grid = [grid1, grid2]
```

This is better, because all of the coils in a `CoilSet` must have the same grid. Before, we were requiring the user to specify all of these individual grids and then throwing away all of them except for the first one. The downside of my new method is this extra logic to account for all of the coils within each coil set.

Reply:
Is this still necessary or can we go back to the more general pytree calls for this?

Reply:
I reverted some of my changes to `_CoilObjective` but I still needed to make some major revisions. It turns out that the `master` branch is not functional with general nested trees (we can't optimize a `MixedCoilSet` that contains other `MixedCoilSet`s). That is another bug technically different from the original intent of this PR, but I think fixed it here anyways. I also updated the tests to cover more cases like this.